File: wampmsgscapnp-go/parsers/hello.go
package parsers

import (
	"bytes"

	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-messages-capnproto/wampmsgscapnp-go/gen"
)

type Hello struct {
	gen *gen.Hello
}

func NewHelloFields(gen *gen.Hello) messages.HelloFields {
	return &Hello{gen: gen}
}

func (h *Hello) Realm() string {
	realm, _ := h.gen.Realm()
	return realm
}

func (h *Hello) AuthID() string {
	authid, _ := h.gen.Authid()
	return authid
}

func (h *Hello) AuthMethods() []string {
	list, err := h.gen.Authmethods()
	if err != nil {
		return nil
	}
	var authmethods []string
	for i := 0; i < list.Len(); i++ {
		authmethod, _ := list.At(i)
		authmethods = append(authmethods, authmethod)
	}
	return authmethods
}

func (h *Hello) AuthExtra() map[string]any {
	return map[string]any{}
}

func (h *Hello) Roles() map[string]any {
	roles := map[string]any{}
	genRoles, err := h.gen.Roles()
	if err != nil {
		return roles
	}

	if r, err := genRoles.Caller(); err == nil {
		roles["caller"] = map[string]any{
			"caller_identification":    r.CallerIdentification(),
			"call_timeout":             r.CallTimeout(),
			"call_canceling":           r.CallCanceling(),
			"progressive_call_results": r.ProgressiveCallResults(),
		}
	}
	if r, err := genRoles.Callee(); err == nil {
		roles["callee"] = map[string]any{
			"caller_identification":      r.CallerIdentification(),
			"call_timeout":               r.CallTimeout(),
			"call_canceling":             r.CallCanceling(),
			"progressive_call_results":   r.ProgressiveCallResults(),
			"pattern_based_registration": r.PatternBasedRegistration(),
			"shared_registration":        r.SharedRegistration(),
		}
	}
	if r, err := genRoles.Publisher(); err == nil {
		roles["publisher"] = map[string]any{
			"publisher_identification":   r.PublisherIdentification(),
			"publisher_exclusion":        r.PublisherExclusion(),
			"subscriber_blackwhite_list": r.SubscriberBlackWhiteListing(),
			"acknowledge_event_received": r.AcknowledgeEventReceived(),
		}
	}
	if r, err := genRoles.Subscriber(); err == nil {
		roles["subscriber"] = map[string]any{
			"publisher_identification":   r.PublisherIdentification(),
			"pattern_based_subscription": r.PatternBasedSubscription(),
			"event_history":              r.EventHistory(),
		}
	}

	return roles
}

func HelloToCapnproto(h *messages.Hello) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	hello, err := gen.NewHello(seg)
	if err != nil {
		return nil, err
	}

	if err := hello.SetRealm(h.Realm()); err != nil {
		return nil, err
	}
	if err := hello.SetAuthid(h.AuthID()); err != nil {
		return nil, err
	}

	methods, err := hello.NewAuthmethods(int32(len(h.AuthMethods())))
	if err != nil {
		return nil, err
	}
	for i, m := range h.AuthMethods() {
		_ = methods.Set(i, m)
	}
	if err := hello.SetAuthmethods(methods); err != nil {
		return nil, err
	}

	roles, err := hello.NewRoles()
	if err != nil {
		return nil, err
	}

	if callerMap, ok := h.Roles()["caller"].(map[string]any); ok {
		caller, err := roles.NewCaller()
		if err == nil {
			if v, ok := callerMap["caller_identification"].(bool); ok {
				caller.SetCallerIdentification(v)
			}
			if v, ok := callerMap["call_timeout"].(bool); ok {
				caller.SetCallTimeout(v)
			}
			if v, ok := callerMap["call_canceling"].(bool); ok {
				caller.SetCallCanceling(v)
			}
			if v, ok := callerMap["progressive_call_results"].(bool); ok {
				caller.SetProgressiveCallResults(v)
			}
			_ = roles.SetCaller(caller)
		}
	}

	if calleeMap, ok := h.Roles()["callee"].(map[string]any); ok {
		callee, err := roles.NewCallee()
		if err == nil {
			if v, ok := calleeMap["caller_identification"].(bool); ok {
				callee.SetCallerIdentification(v)
			}
			if v, ok := calleeMap["call_timeout"].(bool); ok {
				callee.SetCallTimeout(v)
			}
			if v, ok := calleeMap["call_canceling"].(bool); ok {
				callee.SetCallCanceling(v)
			}
			if v, ok := calleeMap["progressive_call_results"].(bool); ok {
				callee.SetProgressiveCallResults(v)
			}
			if v, ok := calleeMap["pattern_based_registration"].(bool); ok {
				callee.SetPatternBasedRegistration(v)
			}
			if v, ok := calleeMap["shared_registration"].(bool); ok {
				callee.SetSharedRegistration(v)
			}
			_ = roles.SetCallee(callee)
		}
	}

	if publisherMap, ok := h.Roles()["publisher"].(map[string]any); ok {
		publisher, err := roles.NewPublisher()
		if err == nil {
			if v, ok := publisherMap["publisher_identification"].(bool); ok {
				publisher.SetPublisherIdentification(v)
			}
			if v, ok := publisherMap["publisher_exclusion"].(bool); ok {
				publisher.SetPublisherExclusion(v)
			}
			if v, ok := publisherMap["subscriber_blackwhite_list"].(bool); ok {
				publisher.SetSubscriberBlackWhiteListing(v)
			}
			if v, ok := publisherMap["acknowledge_event_received"].(bool); ok {
				publisher.SetAcknowledgeEventReceived(v)
			}
			_ = roles.SetPublisher(publisher)
		}
	}

	if subscriberMap, ok := h.Roles()["subscriber"].(map[string]any); ok {
		subscriber, err := roles.NewSubscriber()
		if err == nil {
			if v, ok := subscriberMap["publisher_identification"].(bool); ok {
				subscriber.SetPublisherIdentification(v)
			}
			if v, ok := subscriberMap["pattern_based_subscription"].(bool); ok {
				subscriber.SetPatternBasedSubscription(v)
			}
			if v, ok := subscriberMap["event_history"].(bool); ok {
				subscriber.SetEventHistory(v)
			}
			_ = roles.SetSubscriber(subscriber)
		}
	}

	if err := hello.SetRoles(roles); err != nil {
		return nil, err
	}

	var data bytes.Buffer
	if err := capnp.NewEncoder(&data).Encode(msg); err != nil {
		return nil, err
	}

	byteValue := byte(messages.MessageTypeHello & 0xFF)
	return append([]byte{byteValue}, data.Bytes()...), nil
}

func CapnprotoToHello(data []byte) (*messages.Hello, error) {
	msg, err := capnp.NewDecoder(bytes.NewReader(data)).Decode()
	if err != nil {
		return nil, err
	}

	hello, err := gen.ReadRootHello(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewHelloWithFields(NewHelloFields(&hello)), nil
}

File: python/wamp_msgs_capnp/parsers/hello.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.hello import Hello, IHelloFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "hello.capnp")
hello_capnp = capnp.load(str(module_file))


class HelloFields(IHelloFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def realm(self) -> str:
        return self._gen.realm

    @property
    def authid(self) -> str:
        return self._gen.authid

    @property
    def authmethods(self) -> list[str]:
        return self._gen.authmethods

    @property
    def authextra(self) -> dict[str, Any]:
        extra = {}
        if self._gen.publicKey:
            extra["pubkey"] = self._gen.publicKey

        return extra

    @property
    def roles(self) -> dict[str, Any]:
        roles: dict[str, Any] = {}

        r = self._gen.roles.caller
        roles["caller"] = {
            "caller_identification": r.callerIdentification,
            "call_timeout": r.callTimeout,
            "call_canceling": r.callCanceling,
            "progressive_call_results": r.progressiveCallResults,
        }

        r = self._gen.roles.callee
        roles["callee"] = {
            "caller_identification": r.callerIdentification,
            "call_timeout": r.callTimeout,
            "call_canceling": r.callCanceling,
            "progressive_call_results": r.progressiveCallResults,
            "pattern_based_registration": r.patternBasedRegistration,
            "shared_registration": r.sharedRegistration,
        }

        r = self._gen.roles.publisher
        roles["publisher"] = {
            "publisher_identification": r.publisherIdentification,
            "publisher_exclusion": r.publisherExclusion,
            "acknowledge_event_received": r.acknowledgeEventReceived,
        }

        r = self._gen.roles.subscriber
        roles["subscriber"] = {
            "publisher_identification": r.publisherIdentification,
            "pattern_based_subscription": r.patternBasedSubscription,
        }

        return roles


def hello_to_capnproto(h: Hello) -> bytes:
    hello = hello_capnp.Hello.new_message()
    hello.realm = h.realm
    hello.authid = h.authid

    hello.authmethods = h.authmethods
    public_key = h.authextra.get("pubkey", None)
    if public_key is not None:
        hello.publicKey = public_key

    roles = hello.init("roles")

    if "caller" in h.roles:
        caller = roles.init("caller")
        caller.callerIdentification = h.roles["caller"].get("caller_identification", False)
        caller.callTimeout = h.roles["caller"].get("call_timeout", False)
        caller.callCanceling = h.roles["caller"].get("call_canceling", False)
        caller.progressiveCallResults = h.roles["caller"].get("progressive_call_results", False)

    if "callee" in h.roles:
        callee = roles.init("callee")
        callee.callerIdentification = h.roles["callee"].get("caller_identification", False)
        callee.callTimeout = h.roles["callee"].get("call_timeout", False)
        callee.callCanceling = h.roles["callee"].get("call_canceling", False)
        callee.progressiveCallResults = h.roles["callee"].get("progressive_call_results", False)
        callee.patternBasedRegistration = h.roles["callee"].get("pattern_based_registration", False)
        callee.sharedRegistration = h.roles["callee"].get("shared_registration", False)

    if "publisher" in h.roles:
        publisher = roles.init("publisher")
        publisher.publisherIdentification = h.roles["publisher"].get("publisher_identification", False)
        publisher.publisherExclusion = h.roles["publisher"].get("publisher_exclusion", False)
        publisher.acknowledgeEventReceived = h.roles["publisher"].get("acknowledge_event_received", False)

    if "subscriber" in h.roles:
        subscriber = roles.init("subscriber")
        subscriber.publisherIdentification = h.roles["subscriber"].get("publisher_identification", False)
        subscriber.patternBasedSubscription = h.roles["subscriber"].get("pattern_based_subscription", False)

    hello.roles = roles
    packed_data = hello.to_bytes_packed()

    return helpers.prepend_header(Hello.TYPE, packed_data, b"")


def capnproto_to_hello(data: bytes) -> Hello:
    message_data, _ = helpers.extract_message(data)
    hello_obj = hello_capnp.Hello.from_bytes_packed(message_data)

    return Hello(HelloFields(hello_obj))
